README.md
ZWidget creates a 3D-like visual by stacking widgets one above the other and transforming them adequately.

This package is heavily inspired by [ztext.js](https://bennettfeely.com/ztext/).

Since Text is a Widget, and since everything is a Widget in Flutter, apply your 3D effect to any element of your UI!

It was created for the [Flutter Puzzle Hack](https://flutter.dev/events/puzzle-hack).

## Features

Tilt your widgets and simulate 3D easily.
Here are a few examples with `Text`, `FlutterLogo`, `Ìmage`, `Icon` and `CustomPaint`:

![Various static examples](media/zwidget_various_examples.png)

You can also animate them:

![Demo gif](media/zwidget_animated.gif)

The `example` project includes these demos and a few more.

## Usage

ZWidget includes several constructors depending on the direction where you want your perspective effect to be.

### ZWidget.backwards(), ZWidget.forwards() and ZWidget.bothDirections()

![ZWidget direction constructors](media/zwidget_direction_constructors.png)

You must define at least the `midChild`, which is the base Widget that will be displayed.
Depending on the direction you want the perspective to take place, you may also define the extremities and the widget to display between them and the middle child.

Here is a full example for `ZWidget.bothDirections()`:

``` dart
ZWidget.bothDirections(
  // middle layer
  midChild: Container(
    width: 60,
    height: 60,
    color: Colors.blue,
  ),
  midToBotChild: Container(
    width: 60,
    height: 60,
    color: Colors.blueGrey,
  ),
  // layers between mid and top child
  midToTopChild: Container(
    width: 60,
    height: 60,
    color: Colors.lightBlue,
  ),
  // layer at the bottom of the Stack
  botChild: Container(
    width: 60,
    height: 60,
    color: Colors.grey,
  ),
  // layer at the top of the Stack
  topChild: Container(
    width: 60,
    height: 60,
    color: Colors.lightBlueAccent,
  ),
  // rotation around X axis
  rotationX: pi / 3,
  // rotation around Y axis
  rotationY: -pi / 4,
  // Number of layers. Always odd (increased if even)
  layers: 5,
  // Space between layers
  depth: 32,
)
```

`ZWidget.backwards()` and `ZWidget.forwards()` are similar but they provide only the children parameters adequate to their direction.

The `layers` of a `ZWidget` are visibles in the picture below:

![ZWidget layers](media/layers.png)

There is always an odd number of `layers`. If you define an even number, it will be increased to be an odd number. Increase `depth` to increase their spacing.

`ZDirection` can be `forward`, `backward` or `both`.

Mid layer is the `midChild`. Depending on the `ZDirection`, other layers are displayed above or below the `midChild`.
You can define the `botChild` for the child at the bottom of the Stack and `topChild` for the one at the top.
You can also define Widgets between these extremities and `midChild` with `midToTopChild` and `midToBotChild`.

Use the `alignment` property to set the `Transform` origin.

To effectively see the 3D effect, you must define a rotation to the ZWidget using `rotationX` and `rotationY`.

### ZWidget.builder()

In case you would like to control which Widget is displayed at each layer, you may use `Widget.builder()`.
It can be used to make a gradient effect for instance.

Here are a few examples:

![ZWidget builder examples](media/zwidget_builder.png)



### ZWidget()

While the base `ZWidget` constructor works, it may lead newcomers to define parameters that are not needed.
Therefore, it is now deprecated and you should use the other constructors instead.

## Performance considerations

If ZWidget has 11 layers, it means that 11 Widgets will be drawn. Try to limit the complexity of widgets used at each layer.

For instance, an opaque Image could use simple Containers to simulate a 3D effect:

```dart

final child = ZWidget.forwards(
  midChild: Container(
    color: Colors.black,
    // ...
  ),
  topChild: Image(
    // ...
  ),
  rotationX: pi / 3,
  rotationY: -pi / 4,
  layers: 11,
  depth: 16,
);
```

## All parameters

``` dart
  // The builder function used to build each layer
  final Widget Function(int) builder;
  // Space between layers
  final double depth;
  // Perspective direction
  final ZDirection direction;
  // Rotation around X axis
  final double rotationX;
  // Rotation around Y axis
  final double rotationY;
  // Number of layers. Always odd (increased if set even)
  final int layers;
  // Experimental, used to reverse the X and Y rotations
  final bool reverse;
  // Shows a pink border representing the size taken by this `ZWidget`.
  // Painting may go beyond that border.
  final bool debug;
  // The bigger the perspective, the more part of the widget will seem to be far away.
  final double perspective;
  // Transform origin
  final Alignment? alignment;
    // A fade effect (experimental)
  final bool fade;
```

## ZWidget size

Depending on the transformations and the 3D effect your want to achieve, your widget might be drawn beyond its limits.

The `debug` parameter shows a pink border around your `ZWidget` which indicates its layout size in the widget tree but the painting may go beyond that border.

That's because the `Transform` that we use doesn't change the base size of the widget, it just changes how it is rendered. The layout remains the same, but the painting can change a lot. See the `Transform` [documentation](https://api.flutter.dev/flutter/widgets/Transform-class.html) for more infos.
You might want to cut it to fit in the border with a `ClipRect` for instance.

## How does it work?

ZWidget fills a Stack with `layers` Widgets returned by a `builder` function.
These are then transformed to make the 3D effect.
This transformations is done using a `Transform` widget and a `Matrix4`. If you'd like to do something similar but not with `Widget`, you can use the top-level function `zMatrix4()`.

### zMatrix4

This method returns a `Matrix4` used to perform 3D transformations.
You can use it on anything accepting a `Matrix4` (Widget, Canvas, Path, ...).


#### Widget example

``` dart
Transform(
 transform: zMatrix4(xTilt: pi / 3, yTilt: pi / 4, perspective: 5),
 child: Container(
   color: Colors.red,
   width: 100,
   height: 100,
 ),
);
```
ZWidget uses `zMatrix4()` this way.

#### Canvas example

``` dart
canvas.transform(zMatrix4(xTilt: pi/3).storage);
```

#### Path example

``` dart
Path p = Path()
 ..addOval(Rect.fromCenter(
      center: Offset(size.width / 2, size.height / 2),
      width: size.width / 4,
      height: size.height / 4))
  ..transform(zMatrix4(xTilt: pi / 3).storage);
canvas.drawPath(p, Paint()..color = Colors.amber);
```


## Using this package

- [ZPuzzle](https://play-zpuzzle.web.app)

Contact me if you are using this package and I will add you to the list!


CHANGELOG.md
## [0.0.2]

* **[BREAKING]** ZWidget default constructor is not `const` anymore
* Added `builder` constructor to get more control on each layer
* Added `backwards`, `forwards` and `bothDirections` constructors to simplify usage of ZWidget
* Added `zMatrix4()` as a top-level function to get the `Matrix4` used in `ZWidget`
* More compliant with lint rules

## 0.0.1

* Initial release with basic usage for overlapping Widgets

example/lib/animated_example.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:zwidget/zwidget.dart';

import 'example_custom_painter.dart';

class AnimatedExample extends StatefulWidget {
  const AnimatedExample({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _AnimatedExampleState();
  }
}

class _AnimatedExampleState extends State<AnimatedExample> {
  final _contentKey = GlobalKey(); // declare a global key
  Size? _size;
  double? _statusBarHeight;
  double _rotationX = 0;
  double _rotationY = 0;
  Offset? _mousePosition;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _size = MediaQuery.of(context).size;
      _statusBarHeight = MediaQuery.of(context).viewPadding.top;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Animated ZWidget'),
        leading: const BackButton(),
      ),
      body: Stack(children: [
        Center(
          child: Listener(
            onPointerMove: (PointerMoveEvent event) {
              if (event.down && _size != null) {
                RenderBox box =
                    _contentKey.currentContext!.findRenderObject() as RenderBox;
                Offset targetPosition =
                    box.localToGlobal(box.paintBounds.center);

                final distance = event.position - targetPosition;

                setState(() {
                  _mousePosition = event.position -
                      Offset(0, kToolbarHeight + (_statusBarHeight ?? 0));
                  _rotationY = distance.dx / (_size!.width / 2.0) * 2 * pi;
                  _rotationX = distance.dy / (_size!.height / 2.0) * 2 * pi;
                });
              }
            },
            child: ZWidget.backwards(
              key: _contentKey,
              midChild: ClipRect(
                  child: CustomPaint(
                painter: ExampleCustomPainter(),
                size: const Size(200, 200),
              )),
              rotationX: _rotationX,
              rotationY: _rotationY,
              layers: 11,
              depth: 12,
            ),
          ),
        ),
        if (_mousePosition != null)
          Positioned(
            child: Container(
              width: 30,
              height: 30,
              decoration: const BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.white54,
                  boxShadow: [
                    BoxShadow(color: Colors.black26, spreadRadius: 2)
                  ]),
            ),
            left: _mousePosition!.dx - 15,
            top: _mousePosition!.dy - 15,
          ),
      ]),
    );
  }
}


example/lib/example_custom_painter.dart
import 'package:flutter/material.dart';

class ExampleCustomPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawRect(
        Rect.fromCenter(
            center: Offset(size.width / 2.0, size.height / 2.0),
            width: size.width / 4.0,
            height: size.height / 4.0),
        Paint()..color = Colors.black);

    canvas.drawCircle(
        Offset(size.shortestSide / 2.0, size.shortestSide / 2.0),
        size.shortestSide / 3.0,
        Paint()
          ..color = Colors.blue
          ..style = PaintingStyle.stroke
          ..strokeWidth = 2);

    canvas.drawLine(Offset.zero, size.bottomRight(Offset.zero),
        Paint()..color = Colors.red);

    canvas.drawLine(
        size.bottomLeft(Offset.zero),
        size.topRight(Offset.zero),
        Paint()
          ..color = Colors.red
          ..strokeWidth = 4);

    canvas.drawCircle(Offset(size.shortestSide / 8.0, size.shortestSide / 2.0),
        size.shortestSide / 10.0, Paint()..color = Colors.purple);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:example/animated_example.dart';
import 'package:example/example_custom_painter.dart';
import 'package:flutter/material.dart';
import 'package:zwidget/zwidget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'ZWidget Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: const ZWidgetShowcase(),
    );
  }
}

class ZWidgetShowcase extends StatefulWidget {
  const ZWidgetShowcase({Key? key}) : super(key: key);

  @override
  State<ZWidgetShowcase> createState() => _ZWidgetShowcaseState();
}

class _ZWidgetShowcaseState extends State<ZWidgetShowcase> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('ZWidget Example'),
        actions: [
          IconButton(
            icon: const Icon(Icons.animation),
            onPressed: () {
              Navigator.push(
                  context,
                  MaterialPageRoute(
                      builder: (context) => const AnimatedExample()));
            },
          )
        ],
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            _backwardOnly(),
            _forwardOnly(),
            _bothDirections(),
            _builder(),
            _otherWidgets(),
          ],
        ),
      ),
    );
  }

  Widget _backwardOnly() {
    return Column(children: [
      _title("ZWidget.backwards()"),
      Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        ZWidget.backwards(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.blue,
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            color: Colors.black,
          ),
          rotationX: pi / 3,
          rotationY: -pi / 4,
          layers: 5,
          depth: 8,
        ),
        ZWidget.backwards(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.red,
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            color: Colors.black,
          ),
          rotationX: -pi / 6,
          rotationY: -pi / 4,
          layers: 15,
          depth: 20,
        ),
        ZWidget.backwards(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.green,
            ),
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black,
            ),
          ),
          rotationX: -pi / 3,
          rotationY: 0,
          layers: 5,
          depth: 8,
        ),
        ZWidget.backwards(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.amber,
            ),
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.red,
            ),
          ),
          botChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black,
            ),
          ),
          rotationX: -pi / 6,
          rotationY: pi / 6,
          layers: 11,
          depth: 16,
          perspective: 1,
          alignment: FractionalOffset.center,
          debug: false,
        )
      ]),
    ]);
  }

  Widget _forwardOnly() {
    return Column(children: [
      _title("ZWidget.forwards()"),
      Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        ZWidget.forwards(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.blue,
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            color: Colors.black.withOpacity(0.4),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            color: Colors.blue,
          ),
          rotationX: pi / 3,
          rotationY: -pi / 4,
          layers: 5,
          depth: 16,
        ),
        ZWidget.forwards(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.red.withOpacity(0.6),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            color: Colors.black.withOpacity(0.2),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            color: Colors.red.withOpacity(0.6),
          ),
          rotationX: pi / 3,
          rotationY: -pi / 4,
          layers: 11,
          depth: 16,
        ),
        ZWidget.forwards(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.green,
            ),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black.withOpacity(0.2),
            ),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.greenAccent,
            ),
          ),
          rotationX: -pi / 3,
          rotationY: 0,
          layers: 5,
          depth: 8,
        ),
        ZWidget.forwards(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.red,
            ),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black.withOpacity(0.3),
            ),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.red,
            ),
          ),
          rotationX: -pi / 6,
          rotationY: pi / 6,
          layers: 11,
          depth: 32,
        )
      ]),
    ]);
  }

  Widget _bothDirections() {
    return Column(children: [
      _title("ZWidget.bothDirections()"),
      Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        ZWidget.bothDirections(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.blue,
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            color: Colors.blueGrey,
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            color: Colors.lightBlue,
          ),
          botChild: Container(
            width: 60,
            height: 60,
            color: Colors.grey,
          ),
          topChild: Container(
            width: 60,
            height: 60,
            color: Colors.lightBlueAccent,
          ),
          rotationX: pi / 3,
          rotationY: -pi / 4,
          layers: 5,
          depth: 32,
        ),
        ZWidget.bothDirections(
          midChild: Container(
            width: 60,
            height: 60,
            color: Colors.red,
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            color: Colors.black.withOpacity(0.2),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            color: Colors.white.withOpacity(0.2),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            color: Colors.red,
          ),
          botChild: Container(
            width: 60,
            height: 60,
            color: Colors.red,
          ),
          rotationX: -pi / 6,
          rotationY: -pi / 4,
          layers: 15,
          depth: 24,
        ),
        ZWidget.bothDirections(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.green,
            ),
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black.withOpacity(0.2),
            ),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.lightGreen.withOpacity(0.4),
            ),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.green,
            ),
          ),
          botChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.green,
            ),
          ),
          rotationX: -pi / 3,
          rotationY: 0,
          layers: 9,
          depth: 24,
        ),
        ZWidget.bothDirections(
          midChild: Container(
            width: 60,
            height: 60,
            decoration: const BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.pink,
            ),
          ),
          midToBotChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black.withOpacity(0.3),
            ),
          ),
          midToTopChild: Container(
            width: 60,
            height: 60,
            child: const Center(
                child: Icon(
              Icons.star,
              color: Colors.white,
            )),
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.black.withOpacity(0.3),
            ),
          ),
          topChild: Container(
            width: 60,
            height: 60,
            child: const Center(
                child: Icon(
              Icons.star,
              color: Colors.white,
            )),
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              border: Border.all(width: 2, color: Colors.green),
            ),
          ),
          botChild: Container(
            width: 60,
            height: 60,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: Colors.pink,
              border: Border.all(width: 2, color: Colors.black),
            ),
          ),
          rotationX: -pi / 6,
          rotationY: pi / 6,
          layers: 11,
          depth: 25,
        )
      ]),
    ]);
  }

  Widget _builder() {
    final rainbow = [
      Colors.orange,
      Colors.brown,
      Colors.red,
      Colors.lime,
      Colors.blue,
      Colors.black,
      Colors.green,
      Colors.pink,
      Colors.deepPurpleAccent,
      Colors.cyanAccent,
      Colors.teal,
    ];
    return Column(children: [
      _title("ZWidget.builder()"),
      Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        ZWidget.builder(
          builder: (index) {
            return Container(
              width: 60,
              height: 60,
              color: Colors.blue.withOpacity(0.6),
              child: Center(
                child: Text("$index",
                    style: Theme.of(context).textTheme.headline3),
              ),
            );
          },
          rotationX: pi / 4,
          rotationY: -pi / 4,
          layers: 3,
          depth: 75,
          direction: ZDirection.both,
        ),
        ZWidget.builder(
          builder: (index) {
            return Container(
              width: 60,
              height: 60,
              color: Color.lerp(Colors.amber, Colors.red.shade900, index / 11),
            );
          },
          rotationX: pi / 6,
          rotationY: pi / 5,
          layers: 11,
          depth: 20,
          direction: ZDirection.both,
        ),
        ZWidget.builder(
          builder: (index) {
            return Icon(
              Icons.cake,
              size: 60,
              color: rainbow[index],
            );
          },
          rotationX: pi / 7,
          rotationY: -pi / 4,
          layers: 11,
          depth: 50,
          direction: ZDirection.both,
        ),
        ZWidget.builder(
          builder: (index) {
            return Icon(
              Icons.star,
              size: 60,
              color: Color.lerp(Colors.brown, Colors.grey, index / 7),
            );
          },
          rotationX: -pi / 7,
          rotationY: -pi / 9,
          layers: 11,
          depth: 25,
          direction: ZDirection.both,
        ),
      ]),
    ]);
  }

  Widget _otherWidgets() {
    return Column(children: [
      _title("Other widgets"),
      Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        ZWidget.backwards(
          midChild: const Text(
            'ZWidget',
            style: TextStyle(
                color: Colors.blue, fontSize: 20, fontWeight: FontWeight.bold),
          ),
          midToBotChild: const Text(
            'ZWidget',
            style: TextStyle(
                color: Colors.black, fontSize: 20, fontWeight: FontWeight.bold),
          ),
          rotationX: pi / 4,
          rotationY: pi / 4,
          layers: 5,
          depth: 8,
        ),
        Column(children: [
          ZWidget.backwards(
            midChild: const Text(
              'Example',
              style: TextStyle(
                  color: Colors.red, fontSize: 20, fontWeight: FontWeight.bold),
            ),
            midToBotChild: const Text(
              'Example',
              style: TextStyle(
                  color: Colors.black,
                  fontSize: 20,
                  fontWeight: FontWeight.bold),
            ),
            rotationX: -pi / 3,
            rotationY: 0,
            layers: 5,
            depth: 8,
          ),
          ZWidget.backwards(
            midChild: const Text(
              'Example',
              style: TextStyle(
                  color: Colors.red, fontSize: 20, fontWeight: FontWeight.bold),
            ),
            midToBotChild: const Text(
              'Example',
              style: TextStyle(
                  color: Colors.black,
                  fontSize: 20,
                  fontWeight: FontWeight.bold),
            ),
            rotationX: pi / 3,
            rotationY: 0,
            layers: 5,
            depth: 8,
          ),
        ]),
        ZWidget.backwards(
          midChild: const Text(
            'Text',
            style: TextStyle(
                color: Colors.green, fontSize: 26, fontWeight: FontWeight.bold),
          ),
          midToBotChild: const Text(
            'Text',
            style: TextStyle(
                color: Colors.black, fontSize: 26, fontWeight: FontWeight.bold),
          ),
          rotationX: 0,
          rotationY: -pi / 3,
          layers: 5,
          depth: 8,
        ),
      ]),
      Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: [
          ZWidget.backwards(
            midChild: const FlutterLogo(
              size: 50,
            ),
            rotationX: pi / 6,
            rotationY: pi / 3,
            layers: 11,
            depth: 12,
          ),
          ZWidget.backwards(
            midChild: Image.asset(
              'assets/dashatar.png',
              width: 100,
              height: 100,
              fit: BoxFit.fitHeight,
            ),
            rotationX: pi / 6,
            rotationY: pi / 4,
            layers: 11,
            depth: 12,
          ),
          ZWidget.backwards(
            midChild: const Icon(Icons.home_rounded),
            midToBotChild: const Icon(
              Icons.home_rounded,
              color: Colors.grey,
            ),
            rotationX: pi / 6,
            rotationY: -pi / 4,
            layers: 11,
            depth: 12,
          ),
          ZWidget.backwards(
            midChild: CustomPaint(
              painter: ExampleCustomPainter(),
              size: const Size(50, 50),
            ),
            rotationX: -pi / 3,
            rotationY: -pi / 6,
            layers: 11,
            depth: 12,
          ),
        ],
      ),
    ]);
  }

  Widget _title(String title) {
    return Padding(
      child: Text(
        title,
        style: Theme.of(context).textTheme.bodyText1,
      ),
      padding: const EdgeInsets.only(top: 30, bottom: 20),
    );
  }
}
